### PR TITLE
[flang][clang] Add Visibility specific help text for options

### PR DESCRIPTION
**Pull request by DavidSpickett**
And use it to print the correct default OpenMP version for flang and flang -fc1.

This change adds an optional `HelpTextsForVariants` to options. This allows you to change the help text that gets shown in documentation and `--help` based on the program its being generated for.

As `OptTable` needs to be constexpr compatible, I have used a std::array of help text variants. Each entry is:
(list of visibilities) - > help text string

So for the OpenMP version we have (flang, fc1) -> "OpenMP version for flang is...".

So you can have multiple visibilities use the same string. The number of entries is currently set to 1, and the number of visibilities per entry is 2, because that's the maximum we need for now. The code is written so we can increase these numbers
later, and the unused elements will be initialised.

I have not applied this to group descriptions just because I don't know of one that needs changing. It could easily be enabled for those too if needed. There are minor changes to them just to get it all to compile.

This approach of storing many help strings per option in the 1 driver library seemed preferable to making a whole new library for Flang (even if that would mostly be including stuff from Clang).

### CODE REVIEW

**Comment by DavidSpickett** [comment]
This is a bit of an RFC given that there's only one flang option that requires this that I know of. So this could be too much complication now. Since I was in the area I figured I'd try it anyway, we can always attach this to an issue for later use if necessary.

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -3391,7 +3391,10 @@ def fno_openmp : Flag<["-"], "fno-openmp">, Group<f_Group>,
 def fopenmp_version_EQ : Joined<["-"], "fopenmp-version=">, Group<f_Group>,
   Flags<[NoArgumentUnused]>,
   Visibility<[ClangOption, CC1Option, FlangOption, FC1Option]>,
-  HelpText<"Set OpenMP version (e.g. 45 for OpenMP 4.5, 51 for OpenMP 5.1). Default value is 51 for Clang">;
+  HelpText<"Set OpenMP version (e.g. 45 for OpenMP 4.5, 51 for OpenMP 5.1). Default value is 51 for Clang">,
+  HelpForVisibility<HelpTextForVisibility<FlangOption,
```
Comment (luporl) [inline clang/include/clang/Driver/Options.td:3395]: `-fopenmp-version` is also supported by `flang-new -fc1`.
So shouldn't `FC1Option` also be added to `HelpTextForVisibility` to make `flang-new -fc1 --help` correct?

**Review by luporl — COMMENTED** [review]
I think this can be useful, to make it possible to customize flang help messages, even though only one flang option requires it at the moment.

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (DavidSpickett) [reply clang/include/clang/Driver/Options.td:3395]: Yes, I'll generalise it into a list of alternatives.

**Comment by DavidSpickett** [comment]
@banach-space Your thoughts as the Flang driver owner?

File: clang/include/clang/Driver/Options.td [hunk 2]
```diff
@@ -3382,10 +3382,19 @@ def fopenmp : Flag<["-"], "fopenmp">, Group<f_Group>,
...
+    default, " for ", program);
+}
 def fopenmp_version_EQ : Joined<["-"], "fopenmp-version=">, Group<f_Group>,
   Flags<[NoArgumentUnused]>,
   Visibility<[ClangOption, CC1Option, FlangOption, FC1Option]>,
-  HelpText<"Set OpenMP version (e.g. 45 for OpenMP 4.5, 51 for OpenMP 5.1). Default value is 51 for Clang">;
+  HelpText<OpenMPVersionHelp<"Clang", "51">.str>,
+  HelpTextForVisibilities<[
+    HelpTextForVisibility<FlangOption, OpenMPVersionHelp<"Flang", "11">.str>,
+    HelpTextForVisibility<FC1Option,   OpenMPVersionHelp<"Flang", "11">.str>,
```
Comment (banach-space) [inline clang/include/clang/Driver/Options.td:3396]: I don't envisage the semantics changing depending on whether `flang-new` or `flang-new -fc1` is used. Ideally, this should be just one field with 2 visibilities set.

File: llvm/include/llvm/Option/OptParser.td [hunk 3]
```diff
@@ -93,6 +93,11 @@ class OptionGroup<string name> {
 
 // Define the option class.
 
+class HelpTextForVisibility<OptionVisibility visibility, string text> {
```
Comment (banach-space) [inline llvm/include/llvm/Option/OptParser.td:96]: Naming is hard :) How about `HelpTextVariant`? Basically, `HelpTextForVisibility` refers to the implementation detail ("visibility") and I suggest that the name reflects the functionality provided.

**Review by banach-space — COMMENTED** [review]
+1

I've been thinking about something similar for a while now, but I've never had the cycles to try implementing it. Thanks for working on this! The OpenMP option is an excellent example for "why" to do this. I don't expect _that many_ options requiring this, but IMHO that's secondary. The "help" text should be accurate and this feels like the only way to achieve this. 

While you are here, I would consider looking into `DocBrief`. This one in particular:
* https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Driver/Options.td#L851

That's another place that would benefit from something similar.

I'm a bit short on spare cycles for reviewing stuff :( @luporl , do you have time to go over this? I'll bump this on my list regardless.

**Comment by luporl** [comment]
> I'm a bit short on spare cycles for reviewing stuff :( @luporl , do you have time to go over this? I'll bump this on my list regardless.

@banach-space, yes, I can go over this in more details.

File: clang/include/clang/Driver/Options.td (hunk 2 above)
Comment (DavidSpickett) [reply clang/include/clang/Driver/Options.td:3396]: Sure, I can make it `array<pair<array<visibility>, char*>>`.

Comment (DavidSpickett) [reply clang/include/clang/Driver/Options.td]: This has been done.

File: llvm/include/llvm/Option/OptParser.td (hunk 3 above)
Comment (DavidSpickett) [reply llvm/include/llvm/Option/OptParser.td]: I've changed it to:
HelpTextForVariants (for a single entry in the list, which is many variants to one string)
Help**Texts**ForVariants (for the list itself, which is many HelpTextForVariants)

Which is bound to confuse someone, but it's that or have them be confused why HelpTextForVariant is actually a many to one mapping which seems worse.

File: llvm/utils/TableGen/OptParserEmitter.cpp [hunk 4]
```diff
@@ -312,6 +315,11 @@ static void EmitOptParser(RecordKeeper &Records, raw_ostream &OS) {
     } else
       OS << ", nullptr";
 
+    // Not using Visibility specific text for group help.
```
Comment (DavidSpickett) [inline llvm/utils/TableGen/OptParserEmitter.cpp:318]: This needs to be a loop in case the sizes change.

**Comment by DavidSpickett** [comment]
I've also tested this with increased array sizes. You can check the .inc files to see that they are intialised properly, `<build dir>/lib/ToolDrivers/llvm-lib/Options.inc` is one.

File: llvm/utils/TableGen/OptParserEmitter.cpp [hunk 5]
```diff
@@ -191,6 +191,60 @@ static MarshallingInfo createMarshallingInfo(const Record &R) {
...
+  while (HelpTextsForVariants.size() < MaxVisibilityHelp)
+    HelpTextsForVariants.push_back(
+        {std::vector<std::string>(MaxVisibilityPerHelp, "0"), ""});
+
+  OS << ", (std::array<std::pair<std::array<unsigned, " << MaxVisibilityPerHelp
+     << ">, const char*>, " << MaxVisibilityHelp << ">{{ ";
+
+  assert(HelpTextsForVariants.size() <= MaxVisibilityHelp &&
+         "Too many help text variants to store in "
+         "OptTable::HelpTextsForVariants");
```
Comment (luporl) [inline llvm/utils/TableGen/OptParserEmitter.cpp:215]: This is subjective, but if you agree, I think that moving this assert to the beginning of the function, or before the `OS <<` line would make it easier to read.

File: llvm/utils/TableGen/OptParserEmitter.cpp [hunk 6]
```diff
@@ -191,6 +191,60 @@ static MarshallingInfo createMarshallingInfo(const Record &R) {
...
+
+  for (auto VisibilityHelp = HelpTextsForVariants.cbegin();
+       VisibilityHelp != HelpTextsForVariants.cend(); ++VisibilityHelp) {
```
Comment (luporl) [inline llvm/utils/TableGen/OptParserEmitter.cpp:218]: Nit: save `HelpTextsForVariants.cend()` in a variable.
https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

File: llvm/utils/TableGen/OptParserEmitter.cpp [hunk 7]
```diff
@@ -191,6 +191,60 @@ static MarshallingInfo createMarshallingInfo(const Record &R) {
...
+    auto [Visibilities, Help] = *VisibilityHelp;
+
+    assert(Visibilities.size() <= MaxVisibilityPerHelp &&
+           "Too many visibilities to store in an "
+           "OptTable::HelpTextsForVariants entry");
+    OS << "std::make_pair(std::array<unsigned, " << MaxVisibilityPerHelp
+       << ">{{";
+
+    for (auto Visibility = Visibilities.cbegin();
+         Visibility != Visibilities.cend(); ++Visibility) {
```
Comment (luporl) [inline llvm/utils/TableGen/OptParserEmitter.cpp:228]: Nit: save `Visibilities.cend()` in a variable.

File: llvm/utils/TableGen/OptParserEmitter.cpp [hunk 8]
```diff
@@ -191,6 +191,60 @@ static MarshallingInfo createMarshallingInfo(const Record &R) {
...
+      OS << *Visibility;
+      if (std::next(Visibility) != Visibilities.cend())
+        OS << ", ";
+    }
+
+    OS << "}}, ";
+
+    if (Help.size())
+      OS << "\"" << Help << "\"";
```
Comment (luporl) [inline llvm/utils/TableGen/OptParserEmitter.cpp:237]: If the help text has `"`s in it, will they already be properly escaped?

Comment (DavidSpickett) [reply llvm/utils/TableGen/OptParserEmitter.cpp:237]: No! Good spot, `HelpText` uses `write_cstring` which does so I'll use that.

File: llvm/utils/TableGen/OptParserEmitter.cpp (hunk 5 above)
Comment (DavidSpickett) [reply llvm/utils/TableGen/OptParserEmitter.cpp]: Done.

File: llvm/utils/TableGen/OptParserEmitter.cpp (hunk 7 above)
Comment (DavidSpickett) [reply llvm/utils/TableGen/OptParserEmitter.cpp]: done

File: llvm/utils/TableGen/OptParserEmitter.cpp (hunk 8 above)
Comment (DavidSpickett) [reply llvm/utils/TableGen/OptParserEmitter.cpp]: Done

File: llvm/utils/TableGen/OptParserEmitter.cpp (hunk 6 above)
Comment (DavidSpickett) [reply llvm/utils/TableGen/OptParserEmitter.cpp]: Done.

File: clang/include/clang/Driver/Options.td [hunk 9]
```diff
@@ -3382,10 +3382,18 @@ def fopenmp : Flag<["-"], "fopenmp">, Group<f_Group>,
...
+    default, " for ", program);
+}
 def fopenmp_version_EQ : Joined<["-"], "fopenmp-version=">, Group<f_Group>,
   Flags<[NoArgumentUnused]>,
   Visibility<[ClangOption, CC1Option, FlangOption, FC1Option]>,
-  HelpText<"Set OpenMP version (e.g. 45 for OpenMP 4.5, 51 for OpenMP 5.1). Default value is 51 for Clang">;
+  HelpText<OpenMPVersionHelp<"Clang", "51">.str>,
+  HelpTextsForVariants<[
+    HelpTextVariant<[FlangOption, FC1Option], OpenMPVersionHelp<"Flang", "11">.str>
+  ]>;
```
[the remark this answers is not in the document]
Comment (DavidSpickett) [reply clang/include/clang/Driver/Options.td]: Done.

**Review by luporl — APPROVED** [review]
LGTM, thanks for addressing all my comments!

**Pull request merged**